### DISCUSSION TITLE
Need ProprietaryMessageA example

### DISCUSSION THREAD

**Discussion by Kuzma30** [Open-Agriculture/AgIsoStack-plus-plus#116] [General]
Hello.
Is stack support proprietary message protocol?

**Comment by ad3154** [comment]
Hello there. The short answer is yes, proprietary messaging formats are supported.

The long answer is: yes, but because two or more manufacturers may use the same Proprietary B PGNs, and of course everyone shares the Proprietary A range with completely different message contents, _you have to do the parsing yourself based on the message definition of the message you want to use, and the partner you want to communicate with._ But, the CAN stack will of course deliver to you those messages and let you send those messages with full transport layer support up to the max that the Extended Transport Protocol allows.

I suggest checking out [the tutorial website](https://delgrossoengineering.com/isobus-tutorial/Tutorials/The%20ISOBUS%20Hello%20World.html), as it explains the basics you'll need to do this, and actually even the first example program sends a Proprietary A message to the broadcast address.

For a more practical example though, let's say hypothetically I have a device where if I send it the Proprietary A PGN (just 0xEF00 as the can stack resolves the 00 to the proper destination automatically for you) with some multiplexer, it sends me back some message with content you want to receive. Assuming you've set up your partner control function to match the target device (see the [Adding a Destination Tutorial](https://delgrossoengineering.com/isobus-tutorial/Tutorials/Adding%20a%20Destination.html) for more info on that) and assuming you populate a buffer with the message contents you want to transmit with the Proprietary A message, then all you would need to do is send the message like this:

```
isobus::CANNetworkManager::CANNetwork.send_can_message(0xEF00, messageData, dataLength, sourceECU, destinationECU);
``` 

And that will send the message you constructed.
Likewise, if you register for Proprietary A messages from that partner, you will receive the response. [This part of the tutorial describes how to receive Proprietary A messages](https://delgrossoengineering.com/isobus-tutorial/Tutorials/Receiving%20Messages.html#receiving-destination-specific-messages).

I hope this helps!
